src/control/lib/control/firmware.go

```diff
@@ -0,0 +1,278 @@
+	if !ok {
+		return errors.Errorf("unable to unpack message: %+v", hr.Message)
+	}
+
+	scmResults := make([]*SCMQueryResult, 0, len(pbResp.ScmResults))
+
+	for _, pbScmRes := range pbResp.ScmResults {
+		devResult := &SCMQueryResult{
+			Info: &storage.ScmFirmwareInfo{
+				ActiveVersion:     pbScmRes.ActiveVersion,
+				StagedVersion:     pbScmRes.StagedVersion,
+				ImageMaxSizeBytes: pbScmRes.ImageMaxSizeBytes,
+				UpdateStatus:      storage.ScmFirmwareUpdateStatus(pbScmRes.UpdateStatus),
+			},
+		}
+		if err := convert.Types(pbScmRes.Module, &devResult.Module); err != nil {
+			return errors.Wrapf(err, "unable to convert module")
+		}
+		if pbScmRes.Error != "" {
+			devResult.Error = errors.New(pbScmRes.Error)
+		}
+		scmResults = append(scmResults, devResult)
+	}
+
+	qr.HostSCMFirmware[hr.Addr] = scmResults
+	return nil
+}
+
+// FirmwareQuery concurrently requests device firmware information from
+// all hosts supplied in the request's hostlist, or all configured hosts
+// if not explicitly specified. The function blocks until all results
+// (successful or otherwise) are received, and returns a single response
+// structure containing results for all host firmware query operations.
+func FirmwareQuery(ctx context.Context, rpcClient UnaryInvoker, req *FirmwareQueryReq) (*FirmwareQueryResp, error) {
+	if !req.SCM && !req.NVMe {
+		return nil, errors.New("no device types requested")
+	}
+
+	req.setRPC(func(ctx context.Context, conn *grpc.ClientConn) (proto.Message, error) {
+		return ctlpb.NewMgmtCtlClient(conn).FirmwareQuery(ctx, &ctlpb.FirmwareQueryReq{
+			QueryScm:  req.SCM,
+			QueryNvme: req.NVMe,
+		})
+	})
+
+	unaryResp, err := rpcClient.InvokeUnaryRPC(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := new(FirmwareQueryResp)
+	for _, hostResp := range unaryResp.Responses {
+		if hostResp.Error != nil {
+			if err := resp.addHostError(hostResp.Addr, hostResp.Error); err != nil {
+				return nil, err
+			}
+			continue
+		}
+
+		if err := resp.addHostResponse(hostResp); err != nil {
+			return nil, err
+		}
+	}
+
+	return resp, nil
+}
+
+type (
+	// DeviceType is an enum representing the storage device type.
+	DeviceType uint32
+
+	// FirmwareUpdateReq is a request to update firmware for a specific type
+	// of storage device.
+	FirmwareUpdateReq struct {
+		unaryRequest
+		FirmwarePath string
+		Type         DeviceType
+	}
+
+	// HostSCMUpdateMap maps a host name to a slice of SCM update results.
+	HostSCMUpdateMap map[string][]*SCMUpdateResult
+
+	// SCMUpdateResult represents the results of a firmware update
+	// for a single SCM device.
+	SCMUpdateResult struct {
+		Module storage.ScmModule
+		Error  error
+	}
+
+	// FirmwareUpdateResp returns the results of firmware update operations.
+	FirmwareUpdateResp struct {
+		HostErrorsResp
+		HostSCMResult HostSCMUpdateMap
+	}
+)
+
+const (
+	// DeviceTypeSCM represents SCM modules.
+	DeviceTypeSCM DeviceType = iota
+	// DeviceTypeNVMe represents NVMe SSDs.
+	DeviceTypeNVMe
```

Comment:
It's usually good practice to define an Unknown zero value in case something forgets to explicitly set the device type. Unless you really want the default type to be SCM.

Reply:
Good catch.